src/crimson/admin/admin_socket.cc

```diff
@@ -155,26 +144,23 @@ auto AdminSocket::execute_command(const std::vector<std::string>& cmd,
 				  ceph::bufferlist&& buf)
   -> seastar::future<tell_result_t>
 {
-  return seastar::with_shared(servers_tbl_rwlock,
-			      [cmd, buf=std::move(buf), this]() mutable {
-    auto maybe_parsed = parse_cmd(cmd);
-    if (auto* parsed = std::get_if<parsed_command_t>(&maybe_parsed); parsed) {
-      stringstream os;
-      string desc{parsed->hook.desc};
-      if (!validate_cmd(nullptr, desc, parsed->params, os)) {
-	logger().error("AdminSocket::execute_command: "
-		       "failed to validate '{}': {}", cmd, os.str());
-	ceph::bufferlist out;
-	out.append(os);
-	return seastar::make_ready_future<tell_result_t>(
-          tell_result_t{-EINVAL, "invalid command json", std::move(out)});
-      }
-      return parsed->hook.call(parsed->params, parsed->format, std::move(buf));
-    } else {
-      auto& result = std::get<tell_result_t>(maybe_parsed);
-      return seastar::make_ready_future<tell_result_t>(std::move(result));
+  auto maybe_parsed = parse_cmd(cmd);
+  if (auto* parsed = std::get_if<parsed_command_t>(&maybe_parsed); parsed) {
+    stringstream os;
+    string desc{parsed->hook.desc};
+    if (!validate_cmd(nullptr, desc, parsed->params, os)) {
+      logger().error("AdminSocket::execute_command: "
+                     "failed to validate '{}': {}", cmd, os.str());
+      ceph::bufferlist out;
+      out.append(os);
+      return seastar::make_ready_future<tell_result_t>(
+        tell_result_t{-EINVAL, "invalid command json", std::move(out)});
     }
-  });
+    return parsed->hook.call(parsed->params, parsed->format, std::move(buf));
+  } else {
+    auto& result = std::get<tell_result_t>(maybe_parsed);
+    return seastar::make_ready_future<tell_result_t>(std::move(result));
+  }
 }
```

Comment:
same here. not return future?

Reply:
no, we cannot assume that `hook.call()` is always synchronous .